api/middleware/router.go

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package middleware
+
+import (
+	"context"
+
+	"go.uber.org/yarpc/api/transport"
+)
+
+// Router is a middleware for defining a customized routing experience for procedures
+type Router interface {
+	// Procedures returns the list of procedures that can be called on this router.
+	// Procedures MUST call into router that is passed in.
+	Procedures(transport.Router) []transport.Procedure
+
+	// Choose returns a handlerspec for the given request and transport.
+	// If the Router cannot determine what to call it should call into the router that was
+	// passed in.
+	Choose(context.Context, *transport.Request, transport.Router) (transport.HandlerSpec, error)
+}
+
+// ApplyRouter applies the given Router middleware to
+// the given Router.
+func ApplyRouter(r transport.RouteTable, m Router) transport.RouteTable {
```

Comment:
Consider calling this `ApplyRouteTable` because it operates on a RouteTable, not Router.